azkv/keysource.go

```diff
@@ -0,0 +1,187 @@
+/*
+Package azkv contains an implementation of the go.mozilla.org/sops/keys.MasterKey interface that encrypts and decrypts the
+data key using Azure Key Vault with the Azure Go SDK.
+*/
+package azkv //import "go.mozilla.org/sops/azkv"
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"go.mozilla.org/sops/logging"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/2016-10-01/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/auth"
+	"github.com/sirupsen/logrus"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("AZKV")
+}
+
+// MasterKey is a Azure Key Vault key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	VaultURL string
+	Name     string
+	Version  string
+
+	EncryptedKey string
+	CreationDate time.Time
+}
+
+func newKeyVaultClient() (keyvault.BaseClient, error) {
+	var err error
+	c := keyvault.New()
+	c.Authorizer, err = auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		log.WithError(err).Error("Failed to create Azure authorizer")
+		return c, err
+	}
+
+	return c, nil
+}
+
+// NewMasterKey creates a new MasterKey from an URL, key name and version, setting the creation date to the current date
+func NewMasterKey(vaultURL string, keyName string, keyVersion string) *MasterKey {
+	return &MasterKey{
+		VaultURL:     vaultURL,
+		Name:         keyName,
+		Version:      keyVersion,
+		CreationDate: time.Now().UTC(),
+	}
+}
+
+// MasterKeysFromURLs takes a comma separated list of Azure Key Vault URLs and returns a slice of new MasterKeys for them
+func MasterKeysFromURLs(urls string) []*MasterKey {
+	var keys []*MasterKey
+	if urls == "" {
+		return keys
+	}
+	for _, s := range strings.Split(urls, ",") {
+		keys = append(keys, NewMasterKeyFromURL(s))
+	}
+	return keys
+}
+
+// NewMasterKeyFromResourceID takes an Azure Key Vault key URL and returns a new MasterKey
+// URL format is {vaultUrl}/keys/{key-name}/{key-version}
+func NewMasterKeyFromURL(url string) *MasterKey {
+	k := &MasterKey{}
+	re := regexp.MustCompile("^(https://[^/]+)/keys/([^/]+)/([^/]+)$")
+	parts := re.FindStringSubmatch(url)
+	if parts == nil || len(parts) < 2 {
+		log.Error("No match!")
+		// !?
```

Comment:
None of the others seem to handle errors in their corresponding functionality. This will cause a panic if it cannot extract the parts, is this acceptable or should I refactor it somehow?